packages/base-driver/lib/basedriver/helpers.js

```diff
@@ -66,73 +67,157 @@ async function retrieveHeaders (link) {
+    if (currentModified.getTime() <= lastModified.getTime()) {
+      logger.debug(`The application at ${link} has not been modified since ${lastModified}`);
       return fullPath;
     }
-    if (currentMaxAge && timestamp) {
-      const msLeft = timestamp + currentMaxAge * 1000 - Date.now();
-      if (msLeft > 0) {
-        logger.debug(`The cached application '${path.basename(fullPath)}' will expire in ${msLeft / 1000}s`);
-        return fullPath;
-      }
-      logger.debug(`The cached application '${path.basename(fullPath)}' has expired`);
+    logger.debug(`The application at ${link} has been modified since ${lastModified}`);
+    return refresh();
+  }
+  if (immutable && currentImmutable) {
+    logger.debug(`The application at ${link} is immutable`);
+    return fullPath;
+  }
+  if (currentMaxAge && timestamp) {
+    const msLeft = timestamp + currentMaxAge * 1000 - Date.now();
+    if (msLeft > 0) {
+      logger.debug(`The cached application '${path.basename(fullPath)}' will expire in ${msLeft / 1000}s`);
+      return fullPath;
     }
+    logger.debug(`The cached application '${path.basename(fullPath)}' has expired`);
   }
   return refresh();
 }
 
 function verifyAppExtension (app, supportedAppExtensions) {
-  if (supportedAppExtensions.includes(path.extname(app))) {
+  if (supportedAppExtensions.map(_.toLower).includes(_.toLower(path.extname(app)))) {
     return app;
   }
   throw new Error(`New app path '${app}' did not have ` +
     `${util.pluralize('extension', supportedAppExtensions.length, false)}: ` +
     supportedAppExtensions);
 }
 
-async function configureApp (app, supportedAppExtensions) {
+async function calculateFolderIntegrity (folderPath) {
+  return (await fs.glob('**/*', {cwd: folderPath, strict: false, nosort: true})).length;
+}
+
+async function calculateFileIntegrity (filePath) {
+  return await fs.hash(filePath);
+}
+
+async function isAppIntegrityOk (currentPath, expectedIntegrity = {}) {
+  if (!await fs.exists(currentPath)) {
+    return false;
+  }
+
+  const isDir = (await fs.stat(currentPath)).isDirectory();
+  // Folder integrity check is simple:
+  // Verify the previous amount of files is not greater than the current one.
+  // We don't want to use equality comparison because of an assumption that the OS might
+  // create some unwanted service files/cached inside of that folder or its subfolders.
+  // Ofc, validating the hash sum of each file (or at least of file path) would be much
+  // more precise, but we don't need to be very precise here and also don't want to
+  // overuse RAM and have a performance drop.
+  if (isDir && await calculateFolderIntegrity(currentPath) >= expectedIntegrity?.folder) {
+    return true;
+  }
+  if (!isDir && await calculateFileIntegrity(currentPath) === expectedIntegrity?.file) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * @typedef {Object} ConfigureAppOptions
+ * @property {?Function} onPostProcess Optional function, which should be applied
```

Comment:
```suggestion
 * @property {(obj: Foo) => (Promise<Retval|undefined>|Retval|undefined)} onPostProcess - Optional function, which should be applied
```

where `Foo` is

```js
/**
 * @typedef {Object} Foo
 * @property {boolean} isUrl - Desc of isUrl
 * @property etc
 */
```

likewise a `@typedef` for whatever `Retval` is supposed to be.

Reply:
changed